gems/sorbet-runtime/test/types/casts.rb

```diff
@@ -44,5 +44,27 @@ class CastsTest < Critic::Unit::UnitTest
         end
       end
     end
+
+    describe "T.assert_type with collections" do
+      it "does not do recursive type-checking of arrays with `assert_type!`" do
+        assert_equal([1], T.assert_type!([1], T::Array[String]))
+      end
+
+      it "does do recursive type-checking of arrays with `check_type_recursive!`" do
+        assert_raises(TypeError) do
+          T.check_type_recursive!([1], T::Array[String])
```

Comment:
Yeah seeing this twice, I'm leaning towards at least putting this in `T::Utils` or something, and not putting it on `T` directly, because I think we don't really want to encourage people to depend on this accidentally.

Reply:
I think that for _now_ I'd advocate for `T::Utils`, but it depends on how heavily we think we'll remove this entirely in the future. I think it makes some sense to keep it around, but if we're otherwise pretty sure it'll be removed at some point, then maybe `T::Private` is the right place for it?